File: UIKit and Animations/WWDC 2014/A Look Inside Presentation Controllers (With a case study on alerts and search)/Sample_Codes/LookInside Presentation Controllers, Adaptivity, and Custom Animator Objects/LookInsidePresentationControllersAdaptivityandCustomAnimatorObjects/LookInside/AAPLOverlayTransitioner.m
/*
 Copyright (C) 2014 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 
  AAPLOverlayAnimatedTransitioning and AAPLOverlayTransitioningDelegate implementations.
  
 */

#import "AAPLOverlayTransitioner.h"
#import "AAPLOverlayPresentationController.h"
#import "AAPLCoolPresentationController.h"

#import "AAPLRootViewController.h"

@implementation AAPLOverlayTransitioningDelegate

- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented presentingViewController:(UIViewController *)presenting sourceViewController:(UIViewController *)source
{
    Class presentationControllerClass;
    
    if([source isKindOfClass:[AAPLRootViewController class]] && [(AAPLRootViewController *)source presentationShouldBeAwesome])
    {
        presentationControllerClass = [AAPLCoolPresentationController class];
    }
    else
    {
        presentationControllerClass = [AAPLOverlayPresentationController class];
    }
    
    return [[presentationControllerClass alloc] initWithPresentedViewController:presented presentingViewController:presenting];
}

- (AAPLOverlayAnimatedTransitioning *)animationController
{
    AAPLOverlayAnimatedTransitioning *animationController = [[AAPLOverlayAnimatedTransitioning alloc] init];
    return animationController;
}

- (id <UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source
{
    AAPLOverlayAnimatedTransitioning *animationController = [self animationController];
    [animationController setIsPresentation:YES];
    
    return animationController;
}

- (id <UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed
{
    AAPLOverlayAnimatedTransitioning *animationController = [self animationController];
    [animationController setIsPresentation:NO];
    
    return animationController;
}

@end

@implementation AAPLOverlayAnimatedTransitioning

- (NSTimeInterval)transitionDuration:(id <UIViewControllerContextTransitioning>)transitionContext
{
    return 0.5;
}

- (void)animateTransition:(id <UIViewControllerContextTransitioning>)transitionContext
{
    UIViewController *fromVC = [transitionContext viewControllerForKey:UITransitionContextFromViewControllerKey];
    UIView *fromView = [fromVC view];
    UIViewController *toVC   = [transitionContext viewControllerForKey:UITransitionContextToViewControllerKey];
    UIView *toView = [toVC view];
    
    UIView *containerView = [transitionContext containerView];
    
    BOOL isPresentation = [self isPresentation];

    if(isPresentation)
    {
        [containerView addSubview:toView];
    }
    
    UIViewController *animatingVC = isPresentation ? toVC : fromVC;
    UIView *animatingView = [animatingVC view];
    
    CGRect appearedFrame = [transitionContext finalFrameForViewController:animatingVC];
    CGRect dismissedFrame = appearedFrame;
    dismissedFrame.origin.x += dismissedFrame.size.width;
    
    CGRect initialFrame = isPresentation ? dismissedFrame : appearedFrame;
    CGRect finalFrame = isPresentation ? appearedFrame : dismissedFrame;
    
    [animatingView setFrame:initialFrame];
    
    [UIView animateWithDuration:[self transitionDuration:transitionContext]
                          delay:0
         usingSpringWithDamping:300.0
          initialSpringVelocity:5.0
                        options:UIViewAnimationOptionAllowUserInteraction | UIViewAnimationOptionBeginFromCurrentState
                     animations:^{
                         [animatingView setFrame:finalFrame];
                    }
                     completion:^(BOOL finished){
                         if(![self isPresentation])
                         {
                             [fromView removeFromSuperview];
                         }
                        [transitionContext completeTransition:YES];
                    }];
}

@end

File: UIKit and Animations/WWDC 2014/View Controller Advancements for iOS8/Sample Codes/CustomPresentation/CustomPresentation/AAPLOverlayTransitioner.swift
//
//  AAPLOverlayTransitioner.swift
//  LookInside_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/18/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

class AAPLOverlayTransitioningDelegate: NSObject, UIViewControllerTransitioningDelegate {
	
	var size: CGSize?
	
	var animationController: AAPLOverlayAnimatedTransitioning {
		let animationController = AAPLOverlayAnimatedTransitioning()
		return animationController
	}
	
	func presentationController(forPresented presented: UIViewController, presenting: UIViewController?, source: UIViewController) -> UIPresentationController? {
		let pc = AAPLOverlayPresentationController(presentedViewController: presented, presenting: presenting)
		pc.size = self.size
		return pc
	}

	func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
		let animationController = self.animationController
		animationController.isPresentation = true
		return animationController
	}
	
	func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
		let animationController = self.animationController
		animationController.isPresentation = false
		return animationController
	}
	
}


class AAPLOverlayAnimatedTransitioning: NSObject, UIViewControllerAnimatedTransitioning {
	
	var isPresentation: Bool!

	func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
		return 0.5
	}
	
	func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
		
		let fromVC = transitionContext.viewController(forKey: .from)
		let toVC = transitionContext.viewController(forKey: .to)
		
		let fromView = fromVC?.view
		let toView = toVC?.view
		
		let containerView = transitionContext.containerView
		
		let isPresentation = self.isPresentation!
		
		if isPresentation {
			containerView.addSubview(toView!)
		}
		
		let animatingVC = isPresentation ? toVC : fromVC
		let animatingView = animatingVC?.view
		
		let appearedFrame = transitionContext.finalFrame(for: animatingVC!)
		var dismissedFrame = appearedFrame
		dismissedFrame.origin.y += 2*dismissedFrame.size.height
		
		let initialFrame = isPresentation ? dismissedFrame : appearedFrame;
		let finalFrame = isPresentation ? appearedFrame : dismissedFrame;

		animatingView?.frame = initialFrame
		
		UIView.animate(withDuration: self.transitionDuration(using: transitionContext), delay: 0, usingSpringWithDamping: 300.0, initialSpringVelocity: 5.0, options: [.allowUserInteraction, .beginFromCurrentState], animations: {
			animatingView?.frame = finalFrame
		}) { (finished) in
			
			if !self.isPresentation {
				fromView?.removeFromSuperview()
			}
			
			transitionContext.completeTransition(true)
		}
	}
	
}

